internal/shared/git/git.go

```diff
@@ -0,0 +1,252 @@
+package git
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	gitHttp "github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+
+	"github.com/numaproj-labs/numaplane/api/v1alpha1"
+	controllerconfig "github.com/numaproj-labs/numaplane/internal/controller/config"
+	"github.com/numaproj-labs/numaplane/internal/kubernetes"
+)
+
+// Valid git transports url
+// Source -https://pkg.go.dev/github.com/whilp/git-urls@v1.0.0
+var (
+	// scpSyntax was modified from https://golang.org/src/cmd/go/vcs.go.
+	scpSyntax  = regexp.MustCompile(`^([a-zA-Z0-9-._~]+@)?([a-zA-Z0-9._-]+):([a-zA-Z0-9./._-]+)(?:\?||$)(.*)$`)
+	Transports = NewTransportSet(
+		"ssh",
+		"git",
+		"git+ssh",
+		"http",
+		"https",
+		"ftp",
+		"ftps",
+		"rsync",
+		"file",
+	)
+)
+
+// Checks for valid remote repository
+
+type TransportSet map[string]struct{}
+
+// NewTransportSet returns a TransportSet with the items' keys mapped
+// to empty struct values.
+func NewTransportSet(items ...string) TransportSet {
+	t := make(TransportSet)
+	for _, item := range items {
+		t[item] = struct{}{}
+	}
+	return t
+}
+
+// Valid returns true if transport is a known Git URL scheme and false
+// if not.
+func (t TransportSet) Valid(transport string) bool {
+	_, ok := t[transport] // Directly checking the key in the map
+	return ok
+}
+
+func CheckGitURL(gitURL string) bool {
+	u, err := url.Parse(gitURL)
+	if err == nil && !Transports.Valid(u.Scheme) {
+		return false
+	}
+	return true
+}
+
+// Parser converts a string into a URL.
+type Parser func(string) (*url.URL, error)
```

Comment:
I feel these code can go into a `util` package instead of inside `git`. Can you move it to `util`?

Reply:
Specifically, you can put it under `internal/shared/util/`.

Reply:
This basically parses the GitUrls only so I named it to GitUrlParser

Reply:
I would prefer to put the `GitUrlParser` related logic to a new file like `git_url_parser.go `. But if you like, you can keep if the way it is now. I can do that in the follow up PR.

Reply:
ok in the followup pr i will fix that